destination/src/main/scala/quasar/plugin/sqlserver/destination/Typer.scala

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2020 Precog Data
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.plugin.sqlserver.destination
+
+import quasar.api.{ColumnType, Labeled}
+import quasar.api.push.TypeCoercion
+import quasar.api.push.param._
+import quasar.connector.destination.Constructor
+
+import scala._, Predef._
+
+import cats.data.NonEmptyList
+
+object Typer {
+  import Constructor._
+  import SQLServerType._
+
+  def coerce(scalarType: ColumnType.Scalar): TypeCoercion[SQLServerTypeId] =
+    coercions(scalarType)
+
+  def construct(id: SQLServerTypeId): Either[SQLServerType, Constructor[SQLServerType]] =
+    id match {
+      case tpe: SQLServerTypeId.SelfIdentified => Left(tpe)
+      case hk: SQLServerTypeId.HigherKinded => Right(hk.constructor)
+    }
+
+  def inferScalar(tpe: SQLServerType): Option[ColumnType.Scalar] =
+    coercions.toList
```

Comment:
Wouldn't `collectFirst` work?

Reply:
My intent was to express there must be _exactly_ one match, `collectFirst` doesn't capture that, but there might be a clearer way to implement this.

Reply:
I like this idea and attempted it locally, but for some reason the `upsert delete rows with long typed primary key` example in `SQLServerDestinationSeekSinksSpec` begins failing as a result and I haven't been able to figure out why.

I'm going to merge what is approved and can try to revisit this when I have more time.